### ISSUE TITLE
Add Filters to Event Listerner Components

### ISSUE THREAD

**Issue by mad-moo** [rio-labs/rio#203]
When subscribing to events, e.g. using `rio.PointerEventListener`, the client-side component receives events and passes them on. Python then decides whether it cares about a particular event (e.g. by checking the mouse button) and either performs an action or not.

This is a problem for the client side, because it means that the components cannot tell whether they should eat an event or let it bubble on. This affects

- `rio.PointerEventListener`
- `rio.KeyEventListener`

(And maybe others?)

## Proposed solution

Add some sort of "filter" parameters, that limit the events that components will report to the client side. When a matching event is encountered, it is passed on and eaten. Other events are left unchanged.

**Comment by ricor07** [comment]
Hello, I'd like to work on the project by fixing this issue. Can you assign this to me? Thanks

**Comment by mad-moo** [comment]
Hey! I think this is gonna need some discussion first. Like, how does the user inform Rio which events they're interested in?

If you have an API in mind, can you share your idea?

**Comment by iwr-redmond** [comment]
You may wish to consider adopting some of the ideas from QT, as [psygnal](https://github.com/pyapp-kit/psygnal) and [various](https://github.com/alex-petrenko/signal-slot) [older](https://github.com/dgovil/PySignal) [libraries](https://github.com/complynx/qsignal) have already done.

**Comment by mad-moo** [comment]
@iwr-redmond I don't think that's related? We have event handlers, the question is how to inform them which exact mouse buttons, keys, etc to listen for

**Comment by iwr-redmond** [comment]
If I haven't lost my vertical hold, listeners and events aren't linked, whereas in QT and re-implementations like psygnal they are connected by name. This doesn't require one-to-one connections, per the [QT for Python documentation](https://doc.qt.io/qtforpython-6/tutorials/basictutorial/signals_and_slots.html):

> You can connect as many signals as you want to a single slot, and a signal can be connected to as many slots as you need. 

In psygnal, the design pattern looks like [this](https://github.com/pyapp-kit/psygnal#Usage):

```python

# signal
class MyObject:
    value_changed = Signal(str)

# object
my_obj = MyObject()

# slot
@my_obj.value_changed.connect
def on_change(new_value: str):
    print(f"The value changed to {new_value}!")
```

If you implemented something similar in Rio, I think that your "filter" would be name-based. Optionally, you could make it possible for users to ignore the need to connect events and listeners by name by creating an "all" or "generic" default.

**Comment by Aran-Fey** [comment]
I think the simplest solution would be to pass a dict of `{hotkey: callback_function}` into the constructor, like

```
rio.KeyEventListener(hotkeys={('control', 'f'): self.open_search_bar})
```

**Comment by mad-moo** [comment]
I agree, though I'd go for more readable string keys. Something like `{'ctrl+f': self.open_search_bar}`
